Angular2 feed

magarcia opened the pull request.

ancorgs commented:
This pull request looks quite messy. Some examples:

It includes commit 4ea0954fa4a4bf172ef90a9ca29debd7ee9257b5 which is already in the Angular-2 branch. Even worst. The dummy test that commit is supposed to delete is present in the result of the PR.

Something similar happens with many other commits (like 6227e1883b60d99eda858203ba582da584cdf65e), that are already in our angular-2 branch but are also included in this PR.

The PR adds 6 merges to the history. It's an incremental change on what we currently have in angular-2. It should apply cleanly without all those merges.

Please, try to use rebase, cherry-pick and friends to build a clean PR. Ask for help if you need it.

magarcia commented:
@ancorgs sorry, caos with branches. I think this PR is most clear now.

imobachgs commented on src/app/feed/buttons/audio-button.component.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  public showsAudioOff() {
+    return (!this.feed.isPublisher && !this.feed.isIgnored && !this.feed.getAudioEnabled());
```
You could apply the De Morgan's law here:

```!(this.feed.isPublisher || this.feed.isIgnored || this.feed.getAudioEnabled())```

imobachgs commented on src/app/feed/shared/connection-config.model.ts:
```diff
@@ -0,0 +1,119 @@
+ * of the MIT license.  See the LICENSE.txt file for details.
+ */
+
+import * as _ from "lodash";
```
Maybe `import _ from 'lodash';` would work.

magarcia replied:
For some reason `import _ from "lodash";` not works correctly, but using `import * as _ from "lodash";` solves the problem.

imobachgs commented on src/app/feed/shared/connection-config.model.ts:
```diff
@@ -0,0 +1,119 @@
+import * as _ from "lodash";
+
+export interface IWanted {
+  audio?: boolean;
```
Do we need them to be optional? After all, they're just booleans.

magarcia replied:
In [this method](https://github.com/jangouts/jangouts/pull/147/files#diff-b5c6bf8f85bc96337664b5bdf08b7ac6R100) the requested attribute is set as an empty object, so audio and video need to be optional in the interface. I will put a TODO comment to review this code, but by now I want to keep moving with the migration.

imobachgs commented on src/app/feed/shared/feed-connection.model.ts:
```diff
@@ -0,0 +1,194 @@
+    } else {
+      this.config = new ConnectionConfig(this.pluginHandle, options.values, null, options.ok);
+    }
+  };
```
Remove the semicolon, please.

imobachgs commented on src/app/feed/shared/feed.model.ts:
```diff
@@ -0,0 +1,337 @@
+  /*
+   * Constructor should only require injected parameters, this is the reason to
+   * has setAttrs method. When full app upgraded to Angular 2 DataChannel service
+   * should be injected in the consturctor.
```
typo in *constructor*.

imobachgs commented on src/app/feed/shared/connection-config.model.spec.ts:
```diff
@@ -0,0 +1,151 @@
+    expect(this.pluginHandle.send).toHaveBeenCalled();
+  });
+
+  it("should not be defined until confirm", () => {
```
When you write unit tests, you're testing/describing the public API. So I would use `describe` to make clear what you're testing. For example, here you're testing the `get` method.

```javascript
describe("#get", () => {
  it("...")
});
```

On the other hand, with the current test, it's not clear what's expected from the `get` method.

imobachgs commented on src/app/feed/shared/connection-config.model.spec.ts:
```diff
@@ -0,0 +1,151 @@
+    expect(window.console.error).toHaveBeenCalled();
+  });
+
+  it("should call ok callback when confirm", () => {
```
In this case, for example, you're testing the `confirm` method. So just wrap it in a describe.

magarcia commented:
I updated the tests with your suggestions and added some missing tests for concrete methods in services and components that had not tested. The `feed/buttons` tests still missing, but I'm on it.

imobachgs commented on src/app/feed/feed.component.spec.ts:
```diff
@@ -0,0 +1,81 @@
+
+  describe("#thumbnailTag", () => {
+
+    it("should return placeholder if highlighted", () => {
```
That's not important (just a note), but you can get rid of the _should_ to make the description shorter: "returns placeholder if highlighted".

imobachgs commented on src/app/feed/shared/connection-config.model.spec.ts:
```diff
@@ -41,111 +42,97 @@ describe("Service: ConnectionConfig", () => {
-    spyOn(window.console, "error");
-
-    expect(cconfig.get()).not.toBeDefined();
+    it("should show an error when pluginHandle raise an error", () => {
```
raise -> raises

imobachgs commented on src/app/feed/shared/connection-config.model.ts:
```diff
@@ -7,6 +7,7 @@
 
 import * as _ from "lodash";
 
+// [TODO] - Review if attrs should be optionals (see `configure` method)
```
optionals -> optional

imobachgs commented on src/app/feed/shared/feed-connection.model.spec.ts:
```diff
@@ -30,102 +30,248 @@ describe("Service: FeedConnection", () => {
+      );
+    });
+
+    it("should call success callback when webRTC offer success", () => {
```
success -> succeed

imobachgs replied:
I think this test does not belong to here as you're testing the internals of plugin handle. IMHO it makes sense as a pluginHandle unit test, but not as part of `FeedConnection` model.

On the other hand, you're faking the `createOffer` (and calling `options.success()`), so I think that the original code does not get executed.

The same applies to the test below.

imobachgs commented on src/app/feed/shared/feed-connection.model.spec.ts:
```diff
@@ -30,102 +30,248 @@ describe("Service: FeedConnection", () => {
+      expect(this.options.success).toHaveBeenCalled();
+    });
+
+    it("should call success callback when webRTC offer success", () => {
```
success -> succeeds

imobachgs commented on src/app/feed/shared/feed-connection.model.spec.ts:
```diff
@@ -30,102 +30,248 @@ describe("Service: FeedConnection", () => {
+      );
+    });
+
+    it("should send a request start message when webRTC answer success", () => {
```
As above, I think this test does not cover the original code, right? Testing anonymous callbacks in this case (depending on `pluginHandle` and connections) can be tricky. I think that we should try another approach (de-anonymizing the callbacks, for example).

However, I don't want this issue to block us, so I propose postponing these tests (remove the current ones). Please, add a Trello card to not forget about them.

magarcia replied:
I don't think I'm testing the `pluginHandler`, in these tests the `pluginHandler` it's a fake. In this test I'm faking the `createAnswer` function to test the anonymous callbacks, but this could be a problem if  `pluginHandler` changes his public functions.

De-anonymizing callbacks could be a solution, but IMHO this callback should be private methods and could not be easily testable.

imobachgs commented on src/app/feed/shared/feed-connection.model.spec.ts:
```diff
@@ -30,102 +30,248 @@ describe("Service: FeedConnection", () => {
+      );
+    });
+
+    it("should shows an error when webRTC answer fails", () => {
```
Same as above.

imobachgs commented on src/app/feed/shared/feed-connection.model.spec.ts:
```diff
@@ -30,102 +30,248 @@ describe("Service: FeedConnection", () => {
+      };
+    });
+
+    it("should create a new config if not exist", () => {
```
if it doesn't exist

imobachgs commented on src/app/feed/shared/feed-connection.model.spec.ts:
```diff
@@ -30,102 +30,248 @@ describe("Service: FeedConnection", () => {
+      this.connection.config = jasmine.createSpyObj("config", ["get"]);
+      let result: any = this.connection.getConfig();
+
+      expect(this.connection.config.get).toHaveBeenCalled();
```
You're testing the `config.get` is called but you're not testing whether it's returned. So the following implementation would pass the test:

```javascript
if (this.config) {
  this.config.get();
  return "another-thing";
}
return;
```

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -81,77 +78,70 @@ describe("Service: Feed", () => {
 
-      let result: boolean = feed.isEnabled("audio");
+      let result: boolean = this.feed.isEnabled("audio");
       expect(connection.getConfig).toHaveBeenCalled();
```
I would just say:

```javascript
expect(this.feed.isEnabled("audio")).toBe(connectionStatus.audio)
```

imobachgs replied:
Moreover you can use describes like this (just an sketch to illustrate the general idea):

```javascript
describe("when audio is enabled in the connection", function() {
  let connectionStatus: any = {
    audio: true
  };

  it("returns true", function() {
    expect(this.feed.isEnabled("audio")).toBe(true);
  });
});

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -194,136 +184,279 @@ describe("Service: Feed", () => {
 
-      let result: boolean = feed.getVideoEnabled();
+      let result: boolean = this.feed.getVideoEnabled();
       expect(connection.getConfig).toHaveBeenCalled();
```
Implementation detail that you could ignore.

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -194,136 +184,279 @@ describe("Service: Feed", () => {
-  describe("isConnected", () => {
-    it("should return false if not connection setted", () => {
-      let feed: Feed = new Feed();
+  describe("#igonre", () => {
```
typo: 'ignore'

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -194,136 +184,279 @@ describe("Service: Feed", () => {
+      this.feed.setAttrs({ connection: connection });
 
-      expect(feed.isConnected()).toBe(false);
+      this.feed.disconnect();
```
I think you wanted to call `ignore`, right?

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -194,136 +184,279 @@ describe("Service: Feed", () => {
+
+
+  describe("#stopIgnoring", () => {
+    it("should be connected", () => {
```
"should set the feed as connected" or something like that. What do you think?

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -194,136 +184,279 @@ describe("Service: Feed", () => {
+    });
 
-      expect(feed.waitingForConnection()).toBe(false);
+    it("should return false if feed was speaking in a time given", () => {
```
*given time*

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -194,136 +184,279 @@ describe("Service: Feed", () => {
-      let feed: Feed = new Feed();
-      feed.setAttrs({});
-      feed.disconnect();
+    it("should return true if feed is not speaking in a time given", () => {
```
*given time*

imobachgs commented on src/app/feed/shared/feeds.service.spec.ts:
```diff
@@ -71,8 +74,8 @@ describe("Service: FeedsService", () => {
-  describe("allFeeds",  () => {
-    it("should return and empty array when not feeds added", () => {
+  describe("#allFeeds",  () => {
+    it("should return an empty array when not feeds added", () => {
```
*no feeds were added* or *no feeds are present* or...

imobachgs commented on src/app/feed/shared/feeds.service.spec.ts:
```diff
@@ -209,36 +198,72 @@ describe("Service: FeedsService", () => {
+      let response: Promise<any> = this.feeds.waitFor(1, 3, 10);
 
       tick(10);
       tick(10);
```
I know it's not part of this PR but, why do you call `tick` twice?

magarcia replied:
I don't know why but if I run `tick(20)` once fails and run `tick(10)` twice works. So for each interval cycle is necessary run one tick.

I will add a note above this code to explain this case.

imobachgs commented on src/app/feed/shared/feeds.service.spec.ts:
```diff
@@ -209,36 +198,72 @@ describe("Service: FeedsService", () => {
+
+      expect(this.feeds.find.calls.count()).toBe(4);
+
+      flushMicrotasks(); // resolve promieses
```
typo: *promises*

imobachgs commented on src/app/feed/shared/feeds.service.spec.ts:
```diff
@@ -209,36 +198,72 @@ describe("Service: FeedsService", () => {
+      expect(rejectHandler).toHaveBeenCalled();
     }));
 
+    it("should return a promise that resolve with the asked feed even if it means not wait", <any>fakeAsync((): void => {
```
*not to wait*

imobachgs commented on src/app/feed/shared/speak-observer.service.spec.ts:
```diff
@@ -34,136 +34,142 @@ describe("Service: SpeakObserver", () => {
-  it("should return not speaking when no sound", <any>fakeAsync((): void => {
-    let speakObserver: SpeakObserver = new SpeakObserver({});
+  describe("#isSpeaking", () => {
+    it("should return not speaking when no sound", <any>fakeAsync((): void => {
```
It returns "false", not "not speaking".

imobachgs commented on src/app/feed/shared/speak-observer.service.spec.ts:
```diff
@@ -34,136 +34,142 @@ describe("Service: SpeakObserver", () => {
-    tick(65);
-    tick(65);
-    tick(65);
+    it("should return speaking when sound", <any>fakeAsync((): void => {
```
It should return "true".

imobachgs commented on src/app/feed/shared/speak-observer.service.spec.ts:
```diff
@@ -34,136 +34,142 @@ describe("Service: SpeakObserver", () => {
+      });
 
-    expect(speakObserver.isSpeaking()).toBe(true);
+      // run 3 iterations to detect sound
```
:+1:

imobachgs commented on src/app/feed/shared/speak-observer.service.spec.ts:
```diff
@@ -34,136 +34,142 @@ describe("Service: SpeakObserver", () => {
+    }));
 
-    tick(65);
+    it("should return not speaking when sound stops", <any>fakeAsync((): void => {
```
"not speaking" => "false"

imobachgs commented:
In general it looks good. I only found some issues in the tests but I think they are rather easy to solve. Thanks!

imobachgs commented on src/app/feed/shared/feed-connection.model.ts:
```diff
@@ -104,6 +104,7 @@ export class FeedConnection {
       media.data = false;
     }
 
+    // [TODO] - De-animize the callbacks a tests it
```
*[TODO] de-anonymize callbacks and test them*

imobachgs commented on src/app/feed/shared/feed-connection.model.ts:
```diff
@@ -127,6 +128,8 @@ export class FeedConnection {
    */
   public subscribe(jsepVal: any): void {
+
+    // [TODO] - De-animize the callbacks a tests it
```
Same as above.

imobachgs commented on src/app/feed/shared/feed.model.spec.ts:
```diff
@@ -240,7 +238,7 @@ describe("Model: Feed", () => {
 
   describe("#stopIgnoring", () => {
-    it("should be connected", () => {
+    it("should sets the feed as connected", () => {
```
Sorry! My mistake here: *should set*

imobachgs commented:
Only two typos. It almost LGTM :)

imobachgs commented:
LGTM
